### DISCUSSION TITLE
Testing with dependent providers

### DISCUSSION THREAD

**Discussion by jersonal-com** [rrousselGit/riverpod#1972] [Q&A]
Riverpod is really a fantastic framework. The only thing I am really struggling with is testing. 

I find it particular challenging to implement tests where several async providers depend on one another. A typical example would be a call that first get the uid, then looks up some user data in firebase and fetches an object based on this data from the database. 

When I test this functionality and insert some test data into the mocked Firebase it seems that I have to call each provider in the right order of their dependency before calling the actual widget. Here is a simplified example:

```
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:hooks_riverpod/hooks_riverpod.dart';

final futureIntProvider = FutureProvider<int>((ref) async {
  return Future.value(1);
});

final firstLevelProvider = FutureProvider<int>((ref) async {
  int? value = ref.read(futureIntProvider).value;
  return value!;
});

// Renders the current state and a button that allows incrementing the state
class MyApp extends ConsumerWidget {
  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return MaterialApp(
      home: Consumer(builder: (context, ref, _) {
        final value = ref.read(firstLevelProvider).value ?? 0;
        return ElevatedButton(
          onPressed: () {},
          child: Text('$value'),
        );
      }),
    );
  }
}

void main() {
  testWidgets('Test dependent provider', (tester) async {
    final container = ProviderContainer();
    await container.pump();
    int? result = container.read(futureIntProvider).value;
    print(result); // null
    await container.pump();
    result = container.read(futureIntProvider).value;
    print(result); // 1
    await container.pump();
    result = container.read(firstLevelProvider).value;
    print(result); // null
    await container.pump();
    result = container.read(firstLevelProvider).value;
    print(result); // 1
    await container.pump();

    await tester.pumpWidget(
      UncontrolledProviderScope(
        container: container,
        child: MyApp(),
      ),
    );
    await tester.pumpAndSettle();
    expect(find.text('1'), findsOneWidget);
  });
}
```

Without reading the two providers before calling the widget test results in a failed test. IRL this gets very quickly quite messy and is quite error prone. 

I would appreciate any help if there is a more elegant way to do the "pre fetching" or if I misunderstood a concept and should follow a different pattern for the implementation.

**Comment by rrousselGit** [comment]
Are you looking for `provider.future`?

```dart
final container = ProviderContainer();
await container.read(firstLevelProvider.future);

await tester.pump(...);
```

**Reply by jersonal-com** [reply]
Thanks a lot for the super quick answer.  When reading `provider.future` it is required to first read the `futureIntProvider` and then the `firstLevelProvider`. However, that simplifies the code already quite a bit.

Two quick follow-up questions:
1. For stream providers I assume I would read provider.stream and for a StateNotifierprovider to provider.notifier to "pre-fetch" the value?
2. In order to simplify this procedure - would it make sense to traverse `getAllProviderElementsInOrder()` and that way trigger each provider?

To show the value or firstLevelProvider in the actual app I assume that Riverpod has to go through some dependency management internally that does something very similar. Would it be possible to expose that functionality for testing?
I would assume that using several dependent providers is a common pattern having a way for them to behave in widget testing the same way as in the actual app would be of great value (at least for me - but maybe I am missing something). 

Thanks a lot in advance.

**Reply by rrousselGit** [reply]
> When reading provider.future it is required to first read the futureIntProvider and then the firstLevelProvider. However, that simplifies the code already quite a bit.


No, no need.
Await the provider that your widgets use, and the dependents will be awaited too.

The problem is your implementation of firstLevelProvider. It should instead be:

```dart
final firstLevelProvider = FutureProvider<int>((ref) async {
  final value = await ref.watch(futureIntProvider.future);
  return value;
});
```

**Reply by rrousselGit** [reply]
> For stream providers I assume I would read provider.stream and for a StateNotifierprovider to provider.notifier to "pre-fetch" the value?

Using `.future` there too is fine. This will await for one value to be emitted

**Reply by rrousselGit** [reply]
> In order to simplify this procedure - would it make sense to traverse getAllProviderElementsInOrder() and that way trigger each provider?

Considering my comment about how you should rewrite your firstLevelProvider to use async/await, I don't think this is necessary.

Your suggestion is possible, but it would be probably error-prone, as any custom-made "loading" wouldn't be handled.

**Reply by jersonal-com** [reply]
Thanks for the answer. Super helpful!
